File: src/main/java/keri/projectx/block/BlockFabricator.java
package keri.projectx.block;

import keri.ninetaillib.lib.texture.IIconRegister;
import keri.projectx.ProjectX;
import keri.projectx.client.render.IAnimationHandler;
import keri.projectx.client.render.RenderSimpleGlow;
import keri.projectx.network.ProjectXGuiHandler;
import keri.projectx.tile.TileEntityFabricator;
import keri.projectx.util.EnumXycroniumColor;
import keri.projectx.util.ModPrefs;
import net.minecraft.block.material.Material;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.BlockRenderLayer;
import net.minecraft.util.EnumBlockRenderType;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.registry.GameRegistry;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import javax.annotation.Nullable;

public class BlockFabricator extends BlockProjectX<TileEntityFabricator> implements IAnimationHandler {

    @SideOnly(Side.CLIENT)
    private TextureAtlasSprite[] texture;

    public BlockFabricator() {
        super("fabricator", Material.IRON);
        this.setHardness(1.4F);
    }

    @Nullable
    @Override
    public TileEntityFabricator createNewTileEntity(World world, int meta) {
        return new TileEntityFabricator();
    }

    @Override
    public void registerTileEntities() {
        GameRegistry.registerTileEntity(TileEntityFabricator.class, ModPrefs.MODID + ".tileentity.fabricator");
    }

    @Override
    public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, EntityPlayer player, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ) {
        player.openGui(ProjectX.INSTANCE, ProjectXGuiHandler.GUIID_FABRICATOR, world, pos.getX(), pos.getY(), pos.getZ());
        return true;
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void registerIcons(IIconRegister register) {
        this.texture = new TextureAtlasSprite[3];
        this.texture[0] = register.registerIcon(ModPrefs.MODID + ":blocks/fabricator/fabricator_side");
        this.texture[1] = register.registerIcon(ModPrefs.MODID + ":blocks/fabricator/fabricator_top");
        this.texture[2] = register.registerIcon(ModPrefs.MODID + ":blocks/fabricator/fabricator_bottom");
    }

    @Override
    @SideOnly(Side.CLIENT)
    public TextureAtlasSprite getIcon(int meta, int side) {
        switch(side){
            case 0: return this.texture[2];
            case 1: return this.texture[1];
            case 2: return this.texture[0];
            case 3: return this.texture[0];
            case 4: return this.texture[0];
            case 5: return this.texture[0];
        }

        return null;
    }

    @Override
    @SideOnly(Side.CLIENT)
    public TextureAtlasSprite getAnimationIcon(ItemStack stack, int side) {
        return ProjectX.PROXY.getAnimatedTexture();
    }

    @Override
    @SideOnly(Side.CLIENT)
    public TextureAtlasSprite getAnimationIcon(IBlockAccess world, BlockPos pos, int side) {
        return ProjectX.PROXY.getAnimatedTexture();
    }

    @Override
    @SideOnly(Side.CLIENT)
    public int getAnimationColor(ItemStack stack, int side) {
        return EnumXycroniumColor.BLUE.getColor().rgba();
    }

    @Override
    @SideOnly(Side.CLIENT)
    public int getAnimationColor(IBlockAccess world, BlockPos pos, int side) {
        return EnumXycroniumColor.BLUE.getColor().rgba();
    }

    @Override
    @SideOnly(Side.CLIENT)
    public int getAnimationBrightness(ItemStack stack, int side) {
        return 0x00F000F0;
    }

    @Override
    @SideOnly(Side.CLIENT)
    public int getAnimationBrightness(IBlockAccess world, BlockPos pos, int side) {
        return 0x00F000F0;
    }

    @Override
    @SideOnly(Side.CLIENT)
    public boolean canRenderInLayer(IBlockState state, BlockRenderLayer layer) {
        return layer == BlockRenderLayer.SOLID || layer == BlockRenderLayer.CUTOUT_MIPPED;
    }

    @Override
    @SideOnly(Side.CLIENT)
    public EnumBlockRenderType getRenderType(IBlockState state) {
        return RenderSimpleGlow.RENDER_TYPE;
    }

    @Override
    @SideOnly(Side.CLIENT)
    @SuppressWarnings("deprecation")
    public boolean isOpaqueCube(IBlockState state) {
        return false;
    }

    @Override
    @SideOnly(Side.CLIENT)
    @SuppressWarnings("deprecation")
    public boolean isFullCube(IBlockState state) {
        return false;
    }

}

File: src/main/java/keri/projectx/machine/block/BlockFabricator.scala
package keri.projectx.machine.block

import codechicken.lib.colour.ColourRGBA
import keri.projectx.machine.tile.TileFabricator
import com.projectxy.world.block.BlockXY
import com.projectxy.world.block.glow.ITextureRegisteringAnimationHandler
import com.projectxy.world.{GuiHandler, ProjectXYBlocksTab, ProjectXYWorld, WorldBlocks}
import net.minecraft.block.ITileEntityProvider
import net.minecraft.block.material.Material
import net.minecraft.block.state.IBlockState
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.math.BlockPos
import net.minecraft.util.{EnumFacing, EnumHand}
import net.minecraft.world.{IBlockAccess, World}

class BlockFabricator extends BlockXY(Material.ROCK) with ITextureRegisteringAnimationHandler with ITileEntityProvider with MachineBlock {
  setBlockName("fabricator")
  setCreativeTab(ProjectXYBlocksTab)
  setHardness(1.0F)

  //  override val hasColorMultiplier: Boolean = false
  //
  //  override def getColor(iB4lockAccess: IBlockAccess, x: Int, y: Int, z: Int): Int = color
  override val textureName: String = "fabricator"
  val icons = Array.fill[TextureAtlasSprite](2)(null)

  override def createNewTileEntity(world: World, meta: Int): TileEntity = new TileFabricator


  //  override def getBrightness(world: IBlockAccess, x: Int, y: Int, z: Int): Int = {
  //    world.getTileEntity(x, y, z) match {
  //      case tileFab: TileFabricator =>
  //        tileFab.mode match {
  //          case 0 => if (tileFab.powered) 120 else 220
  //          case 1 => if (tileFab.powered) 220 else 120
  //          case 2 => if (tileFab.powered) 220 else 120
  //          case _ => 220
  //        }
  //      case _ => 220
  //    }
  //  }

  //  override def breakBlock(world: World, x: Int, y: Int, z: Int, block: Block, par6: Int): Unit = {
  //    world.getTileEntity(x, y, z).asInstanceOf[TileFabricator].dropItems()
  //    super.breakBlock(world, x, y, z, block, par6)
  //  }

  override def onBlockActivated(worldIn: World, pos: BlockPos, state: IBlockState, playerIn: EntityPlayer, hand: EnumHand, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): Boolean = {
    if (!playerIn.isSneaking && worldIn.getTileEntity(pos).isInstanceOf[TileFabricator]) {
      playerIn.openGui(ProjectXYWorld, GuiHandler.GuiIds.FABRICATOR.id, worldIn, pos.getX, pos.getY, pos.getZ)
    }
    true
  }

  override def customDrops: Boolean = false

  override def getAnimationColor(world: IBlockAccess, state: IBlockState, pos: BlockPos, side: Int): Int = WorldBlocks.xyColors.head.rgba()

  override def getAnimationColor(stack: ItemStack, side: Int): Int = WorldBlocks.xyColors.head.rgba()
}

object BlockFabricator {
  lazy val color = new ColourRGBA(0, 100, 255, 255).rgb()
}
